File: android/app/src/test/java/io/github/dajiaohuang/solaratlas/SourceIdentityPageTest.java
package io.github.dajiaohuang.solaratlas;

import static org.junit.Assert.*;
import java.io.IOException;
import java.nio.charset.StandardCharsets;
import org.junit.Test;

public class SourceIdentityPageTest {
    private static byte[] bytes(String value) { return value.getBytes(StandardCharsets.UTF_8); }
    private static final String HASH = "a".repeat(64), INVENTORY = "b".repeat(64);
    private static String manifest() { return "{\"apiVersion\":\"solar.api/v1\",\"catalogVersion\":\"test\",\"catalogManifestSha256\":\"" + HASH + "\",\"inventoryManifestSha256\":\"" + INVENTORY + "\"}"; }
    private static String row() { return "{\"id\":\"sb:comet:1P\",\"name\":\"Halley\",\"category\":\"comet\",\"source\":\"synthetic-parser-fixture\",\"sourceRow\":1,\"identityStatus\":\"source-designation\",\"ephemerisStatus\":\"unmapped\"}"; }
    private static String page() { return "{\"apiVersion\":\"solar.api/v1\",\"catalogVersion\":\"test\",\"inventoryManifestSha256\":\"" + INVENTORY + "\",\"sourceRecords\":true,\"identityAssertions\":true,\"uniqueBodySemantics\":\"not-deduplicated\",\"totalRecords\":100,\"limit\":50,\"nextPageToken\":\"next\",\"items\":[" + row() + "]}"; }
    private static SourceIdentityPage decode(String value) throws Exception { return SourceIdentityPage.decode(bytes(value), bytes(manifest()), "https://example.test", "Halley"); }

    @Test public void preservesOriginalSourceIdsAndSeparatePopulationCount() throws Exception {
        SourceIdentityPage result = decode(page()); assertEquals(100, result.totalRecords); assertEquals(1, result.rows.size());
        assertEquals("sb:comet:1P", result.rows.get(0).id); assertEquals("unmapped", result.rows.get(0).ephemerisStatus);
        assertEquals(INVENTORY, result.inventoryHash); assertEquals("Halley", result.query);
        assertThrows(UnsupportedOperationException.class, () -> result.rows.clear());
    }
    @Test public void refusesFalseIdentitySemanticsUnsafeRowsAndUnboundedPages() {
        for (String value : new String[] {page().replace("not-deduplicated", "unique-bodies"), page().replace("\"sourceRecords\":true", "\"sourceRecords\":false"),
                page().replace("\"totalRecords\":100", "\"totalRecords\":0"), page().replace("\"limit\":50", "\"limit\":51"),
                page().replace(row(), row() + "," + row()), page().replace("\"sourceRow\":1", "\"sourceRow\":1.5"),
                page().replace("sb:comet:1P", "bad\\nID"), page().replace(row(), ""),
                page().replace(row(), String.join(",", java.util.Collections.nCopies(51, row()))), page().replace(INVENTORY, HASH)}) {
            assertThrows(StateTileDecoder.ProtocolException.class, () -> decode(value));
        }
    }
    @Test public void rejectsManifestDriftBeforeCursorOrStateUse() throws Exception {
        SourceIdentityPage result = decode(page()); result.requireManifest(bytes(manifest()));
        for (String value : new String[] {manifest().replace(HASH, INVENTORY), manifest().replace(INVENTORY, HASH),
                manifest().replace("test", "new"), manifest().replace("solar.api/v1", "wrong")})
            assertThrows(StateTileDecoder.ProtocolException.class, () -> result.requireManifest(bytes(value)));
    }
    @Test public void validatesBackendQueryAndCancellationBeforeNetwork() throws Exception {
        for (String address : new String[] {"", "http://example.test", "https://user:pass@example.test", "https://example.test?x=1", "https://example.test#x"})
            assertThrows(IOException.class, () -> new SourceIdentityService(address));
        try (SourceIdentityService service = new SourceIdentityService("https://example.test")) {
            assertThrows(IOException.class, () -> service.load("different-query", decode(page())));
            assertThrows(IOException.class, () -> service.load("x".repeat(257), null));
            service.close(); assertThrows(IOException.class, () -> service.load("", null));
        }
    }
}

File: tests/unit/source-identity-page.test.ts
import { describe, expect, it, vi } from 'vitest'
import { inspectSourceIdentityPage, loadSourceIdentityPage, validateSourceIdentityPage } from '../../src/lib/sourceIdentityPage'
import type { StateTileManifest } from '../../src/lib/stateTiles'

const manifest: StateTileManifest = { apiVersion: 'solar.api/v1', catalogVersion: 'test', catalogManifestSha256: 'a'.repeat(64), inventoryManifestSha256: 'b'.repeat(64) }
const row = { id: 'sb:comet:1P', name: 'Halley', category: 'comet', source: 'test-source', sourceRow: 1, identityStatus: 'source-designation', ephemerisStatus: 'not-mapped-to-bundled-kernel', naifId: 1000036 }
function raw() { return { ...manifest, sourceRecords: true, identityAssertions: true, uniqueBodySemantics: 'not-deduplicated', totalRecords: 100, limit: 50, items: [row], nextPageToken: 'next' } }
function json(value: unknown) {
  const body = JSON.stringify(value)
  return new Response(body, { headers: { 'content-type': 'application/json', 'content-length': String(new TextEncoder().encode(body).length) } })
}
const request = () => ({ base: 'https://fixture.invalid', profile: 'full' as const, signal: new AbortController().signal, query: '' })

describe('bounded all-source directory', () => {
  it('preserves a source identity instead of inferring a target or a physical state', () => {
    const page = validateSourceIdentityPage(raw(), manifest)
    expect(page.items[0].id).toBe('sb:comet:1P')
    expect(page.items[0].ephemerisStatus).toBe('not-mapped-to-bundled-kernel')
    expect(page.totalRecords).toBe(100)
    expect(page.items).toHaveLength(1)
    expect(page.items[0]).not.toHaveProperty('states')
  })
  it('rejects false counts, duplicate IDs, unsafe strings, large pages and stale manifest bindings', () => {
    const cases = [
      { totalRecords: -1 }, { totalRecords: 0 }, { limit: 500 }, { sourceRecords: false },
      { uniqueBodySemantics: 'unique-bodies' }, { inventoryManifestSha256: 'c'.repeat(64) },
      { items: [row, row] }, { items: Array(51).fill(row) }, { items: [{ ...row, id: 'bad\nID' }] },
      { items: [{ ...row, sourceRow: 1.5 }] }, { items: [], nextPageToken: 'next' },
    ]
    for (const change of cases) expect(() => validateSourceIdentityPage({ ...raw(), ...change }, manifest)).toThrow()
  })
  it('blocks preview, missing backend and prior cancellation before any request', async () => {
    const fetcher = vi.fn<typeof fetch>()
    await expect(loadSourceIdentityPage({ ...request(), profile: 'preview', fetcher })).rejects.toThrow()
    await expect(loadSourceIdentityPage({ ...request(), base: null, fetcher })).rejects.toThrow()
    await expect(loadSourceIdentityPage({ ...request(), signal: AbortSignal.abort(), fetcher })).rejects.toThrow()
    expect(fetcher).not.toHaveBeenCalled()
  })
  it('fetches one bounded page and pins its cursor to the exact query and dataset', async () => {
    const fetcher = vi.fn<typeof fetch>().mockResolvedValueOnce(json(manifest)).mockResolvedValueOnce(json(raw()))
    const page = await loadSourceIdentityPage({ ...request(), query: 'Halley', fetcher })
    expect(fetcher).toHaveBeenCalledTimes(2)
    expect(String(fetcher.mock.calls[1][0])).toContain('limit=50')
    await expect(loadSourceIdentityPage({ ...request(), previous: page, fetcher })).rejects.toThrow('another query')
    fetcher.mockResolvedValueOnce(json({ ...manifest, inventoryManifestSha256: 'c'.repeat(64) }))
    await expect(loadSourceIdentityPage({ ...request(), query: 'Halley', previous: page, fetcher })).rejects.toThrow('changed')
    expect(fetcher).toHaveBeenCalledTimes(3)
  })
  it('rejects oversized delivery and repeated cursors', async () => {
    const fetcher = vi.fn<typeof fetch>().mockResolvedValueOnce(json(manifest)).mockResolvedValueOnce(new Response('{}', { headers: { 'content-length': String(256 * 1024 + 1) } }))
    await expect(loadSourceIdentityPage({ ...request(), fetcher })).rejects.toThrow('byte budget')
    const page = { ...validateSourceIdentityPage(raw(), manifest), query: '' }
    fetcher.mockResolvedValueOnce(json(manifest)).mockResolvedValueOnce(json(raw()))
    await expect(loadSourceIdentityPage({ ...request(), previous: page, fetcher })).rejects.toThrow('did not advance')
  })
  it('does not issue a state plan after inventory drift, invalid epoch or preview restriction', async () => {
    const page = { ...validateSourceIdentityPage(raw(), manifest), query: '' }
    const fetcher = vi.fn<typeof fetch>().mockResolvedValue(json({ ...manifest, inventoryManifestSha256: 'c'.repeat(64) }))
    await expect(inspectSourceIdentityPage({ ...request(), page, epochTdbJd: NaN, fetcher })).rejects.toThrow()
    await expect(inspectSourceIdentityPage({ ...request(), page, epochTdbJd: 2461287.5, profile: 'preview', fetcher })).rejects.toThrow()
    expect(fetcher).not.toHaveBeenCalled()
    await expect(inspectSourceIdentityPage({ ...request(), page, epochTdbJd: 2461287.5, fetcher })).rejects.toThrow('changed')
    expect(fetcher).toHaveBeenCalledTimes(1)
  })
})

const liveBase = process.env.SOLAR_IDENTITY_API_BASE_URL
it.skipIf(!liveBase)('browses real full-source pages and checks their exact/missing tiles without registry substitution', async () => {
  const params = { ...request(), base: liveBase!, signal: AbortSignal.timeout(90_000) }
  const first = await loadSourceIdentityPage(params)
  expect(first.items).toHaveLength(50)
  const next = await loadSourceIdentityPage({ ...params, previous: first })
  expect(next.items).toHaveLength(50)
  expect(next.items.some(row => first.items.some(prior => prior.id === row.id))).toBe(false)
  const result = await inspectSourceIdentityPage({ ...params, page: first, epochTdbJd: 2461287.5 })
  expect(result.plan.requestIds).toEqual(first.items.map(row => row.id))
  expect(result.plan.exactCount + result.plan.missingCount).toBe(50)
  expect(result.plan.approximateCount).toBe(0)
  expect(result.tiles.flatMap(tile => Array.from({ length: tile.recordCount }, (_, row) => tile.metadata.idAt(row)))).toEqual(result.plan.requestIds)
  console.log(JSON.stringify({ sourceRecords: first.totalRecords, page: 50, exact: result.plan.exactCount, missing: result.plan.missingCount, inventoryHash: first.manifest.inventoryManifestSha256 }))
}, 100_000)
